### PR TITLE
Parse rfc1583 compatibility

### PR DESCRIPTION
**Pull request by corinaminer**
Changes are grouped by commit.

Let me know if I need to add to any other files to make the RFC compatibility info more accessible.

---
This change is [Reviewable](https://reviewable.io/reviews/batfish/batfish/617)

### CODE REVIEW

File: projects/batfish/src/main/java/org/batfish/grammar/cisco/CiscoControlPlaneExtractor.java [hunk 1]
```diff
@@ -1775,6 +1776,10 @@ public void enterS_mac_access_list_extended(S_mac_access_list_extendedContext ct
     _currentMacAccessList = list;
   }
 
+  @Override public void exitRo_rfc1583_compatibility(Ro_rfc1583_compatibilityContext ctx) {
+    _configuration.getCf().setRfc1583Compatible(ctx.NO() == null);
+  }
+
```
Comment (corinaminer) [inline projects/batfish/src/main/java/org/batfish/grammar/cisco/CiscoControlPlaneExtractor.java]: My understanding here is that the method is only called if there's a line about RFC 1583 compatibility, and then ctx.NO() is null iff there's no `no` in that line.

**Comment by dhalperi** [comment]
Reviewed 4 of 8 files at r1.
Review status: 4 of 8 files reviewed at latest revision, 7 unresolved discussions, some commit checks failed.

---

*[projects/batfish-common-protocol/src/main/java/org/batfish/datamodel/vendor_family/cisco/CiscoFamily.java, line 120 at r1](https://reviewable.io:443/reviews/batfish/batfish/617#-KxPMbqf8H5LdOTwJTk0:-KxPMbqf8H5LdOTwJTk1:bgt0qh2) ([raw file](https://github.com/batfish/batfish/blob/193db09d5f8edcfd708f7961f20aa9c6acf9ca6f/projects/batfish-common-protocol/src/main/java/org/batfish/datamodel/vendor_family/cisco/CiscoFamily.java#L120)):*
> ```Java
>   }
> 
>   public Boolean getRfc1583Compatible() { return _rfc1583Compatible; }
> ```

Annotate both the return value and the local variable as `@Nullable`.

---

*[projects/batfish-common-protocol/src/main/java/org/batfish/datamodel/vendor_family/cisco/CiscoFamily.java, line 198 at r1](https://reviewable.io:443/reviews/batfish/batfish/617#-KxPMh9woL3yu7mDVwXl:-KxPMh9woL3yu7mDVwXm:b-43a39b) ([raw file](https://github.com/batfish/batfish/blob/193db09d5f8edcfd708f7961f20aa9c6acf9ca6f/projects/batfish-common-protocol/src/main/java/org/batfish/datamodel/vendor_family/cisco/CiscoFamily.java#L198)):*
> ```Java
>   }
> 
>   public void setRfc1583Compatible(Boolean rfc1583Compatible) { _rfc1583Compatible = rfc1583Compatible; }
> ```

It looks like you have not followed the Batfish instructions and set up the auto formatter. Please do so, and then reformat all java files you've modified.

---

*[projects/batfish/src/main/java/org/batfish/bdp/VirtualRouter.java, line 261 at r1](https://reviewable.io:443/reviews/batfish/batfish/617#-KxPMtMEOnT9STqdF2FH:-KxPMtMEOnT9STqdF2FI:b-98gpz4) ([raw file](https://github.com/batfish/batfish/blob/193db09d5f8edcfd708f7961f20aa9c6acf9ca6f/projects/batfish/src/main/java/org/batfish/bdp/VirtualRouter.java#L261)):*
> ```Java
>      * (see https://www.cisco.com/c/en/us/support/docs/ip/open-shortest-path-first-ospf/7039-1.html#t29)
>      */
>     //TODO: add parsing logic for "(no) compatible rfc1583" command and propagate it to useMin
> ```

Do this, and delete the TODO?

---

*[projects/batfish/src/main/java/org/batfish/bdp/VirtualRouter.java, line 294 at r1](https://reviewable.io:443/reviews/batfish/batfish/617#-KxPN-cta9kB_noGRw60:-KxPN-cta9kB_noGRw61:b-yvrch9) ([raw file](https://github.com/batfish/batfish/blob/193db09d5f8edcfd708f7961f20aa9c6acf9ca6f/projects/batfish/src/main/java/org/batfish/bdp/VirtualRouter.java#L294)):*
> ```Java
>         // Compute the metric from any possible contributing routes, use older RFC by default
>         // as it seems consistent with the GNS3 simulations
>         Boolean useMin = _c.getVendorFamily().getCisco().getRfc1583Compatible();
> ```

This doesn't seem like the right part of the model to have this flag -- it's a per-OSPF process flag. E.g., here's a recommended config; can you support this?

```
router ospf 1
  no compatible rfc1583
!
router ospf 2
  compatible rfc1583
!
```

---

*[projects/batfish/src/test/resources/org/batfish/grammar/cisco/testconfigs/rfc1583Compatible, line 6 at r1](https://reviewable.io:443/reviews/batfish/batfish/617#-KxPP7SqmIRc3XBnIos0:-KxPP7SqmIRc3XBnIos1:bmo407m) ([raw file](https://github.com/batfish/batfish/blob/193db09d5f8edcfd708f7961f20aa9c6acf9ca6f/projects/batfish/src/test/resources/org/batfish/grammar/cisco/testconfigs/rfc1583Compatible#L6)):*
> ```
> router ospf 1
>  compatible rfc1583
> !
> ```

Make this 1 file with 2 OSPF routers and verify that we can set compatibility separately for the two of them.

---

*[projects/batfish/src/test/resources/org/batfish/grammar/cisco/testconfigs/rfc1583NoCompatible, line 5 at r1](https://reviewable.io:443/reviews/batfish/batfish/617#-KxPPDZa0buPqcF1MOmO:-KxPPDZa0buPqcF1MOmP:biivwrb) ([raw file](https://github.com/batfish/batfish/blob/193db09d5f8edcfd708f7961f20aa9c6acf9ca6f/projects/batfish/src/test/resources/org/batfish/grammar/cisco/testconfigs/rfc1583NoCompatible#L5)):*
> ```
> !
> router ospf 1
>  no compatible rfc1583
> ```

(can delete this file now)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/batfish/batfish/617)*

**Comment by dhalperi** [comment]
Reviewed 10 of 10 files at r2.
Review status: 3 of 11 files reviewed at latest revision, 5 unresolved discussions.

---

*[projects/batfish-common-protocol/src/main/java/org/batfish/datamodel/OspfProcess.java, line 73 at r2](https://reviewable.io:443/reviews/batfish/batfish/617#-KxU9bMk0_S2Ig0FM4-D:-KxU9bMk0_S2Ig0FM4-E:bcp0171) ([raw file](https://github.com/batfish/batfish/blob/f5c15a31f10e68148ea3528cd5b95ac578a32d38/projects/batfish-common-protocol/src/main/java/org/batfish/datamodel/OspfProcess.java#L73)):*
> ```Java
>   }
> 
>   public @Nullable Boolean getRfc1583Compatible() { return _rfc1583Compatible; }
> ```

reformat file

---

*[projects/batfish/src/main/java/org/batfish/bdp/VirtualRouter.java, line 294 at r1](https://reviewable.io:443/reviews/batfish/batfish/617#-KxPHzhr-0fSUu2KRizM-r1-294:-KxUAXTeDpHOE5woRuKW:bb6un1j) ([raw file](https://github.com/batfish/batfish/blob/193db09d5f8edcfd708f7961f20aa9c6acf9ca6f/projects/batfish/src/main/java/org/batfish/bdp/VirtualRouter.java#L294)):*
<details><summary><i>Previously, progwriter (Victor Heorhiadi) wrote…</i></summary><blockquote>

At this point we should rely only on vendor-independent config. I believe there should be a toVendorIndependentConfiguration() method that will need to be changed to keep track of this var. Then it can be propagated to the Vrf & Ospf process
</blockquote></details>

Done

---

*[projects/batfish/src/main/java/org/batfish/bdp/VirtualRouter.java, line 294 at r1](https://reviewable.io:443/reviews/batfish/batfish/617#-KxPN-cta9kB_noGRw60:-KxUAbo2JKIA4VKYnzER:b-95hpws) ([raw file](https://github.com/batfish/batfish/blob/193db09d5f8edcfd708f7961f20aa9c6acf9ca6f/projects/batfish/src/main/java/org/batfish/bdp/VirtualRouter.java#L294)):*
<details><summary><i>Previously, dhalperi (Dan Halperin) wrote…</i></summary><blockquote>

This doesn't seem like the right part of the model to have this flag -- it's a per-OSPF process flag. E.g., here's a recommended config; can you support this?

```
router ospf 1
  no compatible rfc1583
!
router ospf 2
  compatible rfc1583
!
```
</blockquote></details>

Ignore - as Batfish doesn't currently support multiple ospf processes, multiple files is better.

---

*[projects/batfish/src/main/java/org/batfish/bdp/VirtualRouter.java, line 271 at r2](https://reviewable.io:443/reviews/batfish/batfish/617#-KxU9kJ4qdeRZnlRO5Ra:-KxU9kJ4qdeRZnlRO5Rb:betnvre) ([raw file](https://github.com/batfish/batfish/blob/f5c15a31f10e68148ea3528cd5b95ac578a32d38/projects/batfish/src/main/java/org/batfish/bdp/VirtualRouter.java#L271)):*
> ```Java
> 
>     // Determine whether to use min based on RFC 1583 compatibility setting
>     Boolean useMin = proc.getRfc1583Compatible();
> ```

simpler:

```java
boolean useMin = MoreObjects.firstNonNull(proc.getRfc1583Compatible(), true);
```

and I'd probably add a comment that contains some of the comments from the other note, perhaps along the lines of:

```java
// Between the original RFC1583 and the newer RFC2328, the OSPF summary metric
// was changed from min to max. Routers (Cisco, Juniper at least) default to RFC1583
// unless explicitly disabled.
```

---

*[projects/batfish/src/main/java/org/batfish/grammar/cisco/CiscoControlPlaneExtractor.java, line 1779 at r2](https://reviewable.io:443/reviews/batfish/batfish/617#-KxUAtB5Pdpd_Iuy1fO_:-KxUAtB5Pdpd_Iuy1fOa:b-l5gclu) ([raw file](https://github.com/batfish/batfish/blob/f5c15a31f10e68148ea3528cd5b95ac578a32d38/projects/batfish/src/main/java/org/batfish/grammar/cisco/CiscoControlPlaneExtractor.java#L1779)):*
> ```Java
>   }
> 
>   @Override public void exitRo_rfc1583_compatibility(Ro_rfc1583_compatibilityContext ctx) {
> ```

auto-format here, too.

---

*[projects/batfish/src/main/java/org/batfish/representation/cisco/OspfProcess.java, line 155 at r2](https://reviewable.io:443/reviews/batfish/batfish/617#-KxUB405RN9Mn0Ifd1Fk:-KxUB405RN9Mn0Ifd1Fl:b-kygpg5) ([raw file](https://github.com/batfish/batfish/blob/f5c15a31f10e68148ea3528cd5b95ac578a32d38/projects/batfish/src/main/java/org/batfish/representation/cisco/OspfProcess.java#L155)):*
> ```Java
>   }
> 
>   public @Nullable Boolean getRfc1583Compatible() { return _rfc1583Compatible; }
> ```

autoformat

---

*[projects/batfish/src/test/java/org/batfish/grammar/cisco/CiscoGrammarTest.java, line 207 at r2](https://reviewable.io:443/reviews/batfish/batfish/617#-KxUBDZnVsZlak8kuruk:-KxUBDZnVsZlak8kurul:be846z6) ([raw file](https://github.com/batfish/batfish/blob/f5c15a31f10e68148ea3528cd5b95ac578a32d38/projects/batfish/src/test/java/org/batfish/grammar/cisco/CiscoGrammarTest.java#L207)):*
> ```Java
>         BatfishTestUtils.getBatfishFromTestrigText(
>             configurationTextMap,
>             Collections.emptySortedMap(),
> ```

autoformat, but LGTM

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/batfish/batfish/617)*

**Comment by corinaminer** [comment]
Review status: 3 of 11 files reviewed at latest revision, 5 unresolved discussions.

---

*[projects/batfish/src/main/java/org/batfish/bdp/VirtualRouter.java, line 294 at r1](https://reviewable.io:443/reviews/batfish/batfish/617#-KxPN-cta9kB_noGRw60:-KxUAS7yPU8rwCw-mghV:bkc62u5) ([raw file](https://github.com/batfish/batfish/blob/193db09d5f8edcfd708f7961f20aa9c6acf9ca6f/projects/batfish/src/main/java/org/batfish/bdp/VirtualRouter.java#L294)):*
<details><summary><i>Previously, dhalperi (Dan Halperin) wrote…</i></summary><blockquote>

This doesn't seem like the right part of the model to have this flag -- it's a per-OSPF process flag. E.g., here's a recommended config; can you support this?

```
router ospf 1
  no compatible rfc1583
!
router ospf 2
  compatible rfc1583
!
```
</blockquote></details>

No, can't support that. Moved the flag down into OspfProcess, though.

---

*[projects/batfish-common-protocol/src/main/java/org/batfish/datamodel/vendor_family/cisco/CiscoFamily.java, line 120 at r1](https://reviewable.io:443/reviews/batfish/batfish/617#-KxPMbqf8H5LdOTwJTk0:-KxUB3w00bp8yJ7xvcl-:b-vzs9fs) ([raw file](https://github.com/batfish/batfish/blob/193db09d5f8edcfd708f7961f20aa9c6acf9ca6f/projects/batfish-common-protocol/src/main/java/org/batfish/datamodel/vendor_family/cisco/CiscoFamily.java#L120)):*
<details><summary><i>Previously, dhalperi (Dan Halperin) wrote…</i></summary><blockquote>

Annotate both the return value and the local variable as `@Nullable`.
</blockquote></details>

Should the argument of the setter be marked `@Nullable` too?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/batfish/batfish/617)*

**Comment by dhalperi** [comment]
Reviewed 8 of 8 files at r3.
Review status: all files reviewed at latest revision, 5 unresolved discussions.

---

*[projects/batfish-common-protocol/src/main/java/org/batfish/datamodel/vendor_family/cisco/CiscoFamily.java, line 120 at r1](https://reviewable.io:443/reviews/batfish/batfish/617#-KxPMbqf8H5LdOTwJTk0:-KxUK4CUM7uVe_NlRXfe:bkfw7bq) ([raw file](https://github.com/batfish/batfish/blob/193db09d5f8edcfd708f7961f20aa9c6acf9ca6f/projects/batfish-common-protocol/src/main/java/org/batfish/datamodel/vendor_family/cisco/CiscoFamily.java#L120)):*
<details><summary><i>Previously, corinaminer (Corina Miner) wrote…</i></summary><blockquote>

Should the argument of the setter be marked `@Nullable` too?
</blockquote></details>

Yes, probably. Good point.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/batfish/batfish/617)*

**Comment by dhalperi** [comment]
:lgtm: after comments are addressed

---

Review status: all files reviewed at latest revision, 5 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/batfish/batfish/617#-:-KxUKAlo6vPzykjVVCjp:b-o64p82)*

**Comment by corinaminer** [comment]
Review status: all files reviewed at latest revision, 6 unresolved discussions.

---

*[projects/batfish/src/main/java/org/batfish/bdp/VirtualRouter.java, line 271 at r2](https://reviewable.io:443/reviews/batfish/batfish/617#-KxU9kJ4qdeRZnlRO5Ra:-KxUKkC5QDmUsTn1Ejb1:b-896fix) ([raw file](https://github.com/batfish/batfish/blob/f5c15a31f10e68148ea3528cd5b95ac578a32d38/projects/batfish/src/main/java/org/batfish/bdp/VirtualRouter.java#L271)):*
<details><summary><i>Previously, dhalperi (Dan Halperin) wrote…</i></summary><blockquote>

simpler:

```java
boolean useMin = MoreObjects.firstNonNull(proc.getRfc1583Compatible(), true);
```

and I'd probably add a comment that contains some of the comments from the other note, perhaps along the lines of:

```java
// Between the original RFC1583 and the newer RFC2328, the OSPF summary metric
// was changed from min to max. Routers (Cisco, Juniper at least) default to RFC1583
// unless explicitly disabled.
```
</blockquote></details>

Done.

---

*[projects/batfish/src/main/java/org/batfish/grammar/cisco/CiscoControlPlaneExtractor.java, line 1779 at r2](https://reviewable.io:443/reviews/batfish/batfish/617#-KxUAtB5Pdpd_Iuy1fO_:-KxUKYmy-qFg3tj3Xq74:b-896fix) ([raw file](https://github.com/batfish/batfish/blob/f5c15a31f10e68148ea3528cd5b95ac578a32d38/projects/batfish/src/main/java/org/batfish/grammar/cisco/CiscoControlPlaneExtractor.java#L1779)):*
<details><summary><i>Previously, dhalperi (Dan Halperin) wrote…</i></summary><blockquote>

auto-format here, too.
</blockquote></details>

Done.

---

*[projects/batfish/src/main/java/org/batfish/representation/cisco/OspfProcess.java, line 155 at r2](https://reviewable.io:443/reviews/batfish/batfish/617#-KxUB405RN9Mn0Ifd1Fk:-KxUKceyNro3PPTc3O3O:b-896fix) ([raw file](https://github.com/batfish/batfish/blob/f5c15a31f10e68148ea3528cd5b95ac578a32d38/projects/batfish/src/main/java/org/batfish/representation/cisco/OspfProcess.java#L155)):*
<details><summary><i>Previously, dhalperi (Dan Halperin) wrote…</i></summary><blockquote>

autoformat
</blockquote></details>

Done.

---

*[projects/batfish/src/test/java/org/batfish/grammar/cisco/CiscoGrammarTest.java, line 207 at r2](https://reviewable.io:443/reviews/batfish/batfish/617#-KxUBDZnVsZlak8kuruk:-KxUKdT675MP5-sKHod0:b-896fix) ([raw file](https://github.com/batfish/batfish/blob/f5c15a31f10e68148ea3528cd5b95ac578a32d38/projects/batfish/src/test/java/org/batfish/grammar/cisco/CiscoGrammarTest.java#L207)):*
<details><summary><i>Previously, dhalperi (Dan Halperin) wrote…</i></summary><blockquote>

autoformat, but LGTM
</blockquote></details>

Done.

---

*[projects/batfish-common-protocol/src/main/java/org/batfish/datamodel/vendor_family/cisco/CiscoFamily.java, line 120 at r1](https://reviewable.io:443/reviews/batfish/batfish/617#-KxPMbqf8H5LdOTwJTk0:-KxUKfwvu3CwJ7ZWzxuW:b-896fix) ([raw file](https://github.com/batfish/batfish/blob/193db09d5f8edcfd708f7961f20aa9c6acf9ca6f/projects/batfish-common-protocol/src/main/java/org/batfish/datamodel/vendor_family/cisco/CiscoFamily.java#L120)):*
<details><summary><i>Previously, dhalperi (Dan Halperin) wrote…</i></summary><blockquote>

Yes, probably. Good point.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/batfish/batfish/617)*

**Comment by dhalperi** [comment]
:lgtm:

---

Reviewed 2 of 2 files at r4.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/batfish/batfish/617#-:-KxUMJQAZHw-u0hnStqr:bnfp4nl)*

**Pull request merged**